### ISSUE TITLE
NXentry names not compliant with NeXus

### ISSUE THREAD

**Issue by prjemian** [silx-kit/silx#2691]
The names for the `NXentry` groups in the HDF5 files, while valid HDF5 item names, are not compliant with the [NeXus naming convention](http://download.nexusformat.org/doc/html/datarules.html#naming-conventions).  For example, given a SPEC file with scans `#S 1` and `#S 2`, it would be converted to this HDF5 structure (omitting attributes and data sets for clarity).

```
<filename>.h5:NXroot
   1.1:NXentry
      instrument:NXinstrument
      measurement:NXcollection
   2.1:NXentry
      instrument:NXinstrument
      measurement:NXcollection
```

The `1.1` and `2.1` refer to the two SPEC scans (with the first header).  While these are valid HDF5 item names, they are not compliant with [NeXus](http://download.nexusformat.org/doc/html/datarules.html#naming-conventions) naming conventions.  Could be fixed by changing the `.` for a compliant character (such as `_`) and adding a prefix character (such as `S`).  The suggested revisions would appear:

```
<filename>.h5:NXroot
   S1_1:NXentry
      instrument:NXinstrument
      measurement:NXcollection
   S2_1:NXentry
      instrument:NXinstrument
      measurement:NXcollection
```

This issue was discovered while [reviewing the *silx* package](https://github.com/prjemian/spec2nexus/issues/141#issuecomment-515782804).

**Comment by vasole** [comment]
Unless I am mistaken, ``recommended`` and ``mandatory`` do not carry the same meaning.

**Comment by prjemian** [comment]
@vasole, thanks for your comment.  From the [NeXus manual]():

> The names of NeXus group and field items must only contain a restricted set of characters. 

The suggested change is compliant with the regular expression: `[A-Za-z_][\w_]*`.

**Comment by vasole** [comment]
The recommended word is used below:

Regular expression pattern for NXDL group and field names

It is recommended that all group and field names contain only these characters:

    lower case letters
    digits
    “_” (underscore character)

and that they begin with a lower case letter.

**Comment by prjemian** [comment]
Following that text is the regular expression I cited.  Names that that do not match either of the regular expressions are not compliant with the NeXus naming convention.  The names presented in *silx* do not match either of the regular expressions.

**Comment by prjemian** [comment]
@vasole: This would be more clear if the related [NeXus issue](https://github.com/nexusformat/definitions/issues/544) were to be completed.

**Comment by vasole** [comment]
I do not discuss the validity of the regular expression.

What I am saying is that the word recommended should then be replaced in that documentation if it is an actual rule and not a recommendation

I always considered NeXus mostly attribute based concerning the groups giving freedom to the group names while providing recommendations (lower_case_with_underscores)

It is quite common to name the NXinstrument group with the name of the beamline and rarely that one is lower case (I have seen SOLEIL data with DIFFABS as name of the NXinstruments group). If the recommendation is not such and it is a rule, then it would not be acceptable.

In this particular case, being the users used to use 1.1, 2.1 and so on when using the SpecFile library I see little advantage switching to 1_1, 2_1 but I will consult with the colleagues/scientists to figure out.

**Comment by prjemian** [comment]
I agree with you about the names.  It was an effort to get the NIAC to agree to upper case in the naming convention.  The older history was the intent that these names would be valid program vairables.  Why is that relevant today?  Usually, they are used as keys in a dictionary. I'll work on the [NeXus issue](https://github.com/nexusformat/definitions/issues/544) next.
